File: dump-cn/457.circular-array-loop.cs
public class Solution {
    public bool CircularArrayLoop(int[] nums)
    {
        int m = nums.Length;
        for (int i = 0; i < m; ++i)
        {
            int ptr1 = i;
            int ptr2 = i;
            int count = 0;
            int first = nums[ptr1];
            do
            {
                if (first * nums[ptr1] < 0)
                {
                    count = 0;
                    break;
                }
                ptr1 = ((ptr1 + nums[ptr1]) % m + m) % m;
                if (first * nums[ptr2] < 0)
                {
                    count = 0;
                    break;
                }
                ptr2 = ((ptr2 + nums[ptr2]) % m + m) % m;
                if (first * nums[ptr2] < 0)
                {
                    count = 0;
                    break;
                }
                ptr2 = ((ptr2 + nums[ptr2]) % m + m) % m;
                count++;
            } while (ptr1 != ptr2);
            if (count > 1 && ((ptr1 + nums[ptr1]) % m + m) % m != ptr1)
                return true;
        }
        return false;
    }
}

File: dump-cn/457.circular-array-loop.java
class Solution {
    int[] flag;
    public boolean circularArrayLoop(int[] nums) {
        flag = new int[nums.length];
        for(int i = 0; i < nums.length; i++) {
        	if(flag[i] != 1) {
        		if(to(i,i,nums)) {
        			return true;
        		}
        	}
        }
        return false;
    }
    public boolean to(int index,int first,int[] nums) {  
        //System.out.println(index);
    	if(flag[index] == 1)
    		return false;
        
    	int i =index+nums[index];
        while(i<0){
            i+=nums.length;
        }
        i%=nums.length;
    	

    	if(nums[i]*nums[index]<0) {
    		return false;
    	}
        if(index == i){
            return false;
        }
    	if(i == first) {
    		return true;
    	}
        if(flag[i] == 2)
            return true;
        flag[index] = 2;
    	boolean p = to(i,first,nums);
        flag[index] = 1;        
        return p;
    }
}